### PR TITLE
feat(teams-limit): Add snackbar to usage stats enabled notification

### LINKED ISSUE #7383: Teams limit: Trigger usage stats
When tariff limits exceeded lets enable Usage stats for billing leaders, so they can check what is going on

Previous issue:
#7379

Designs: https://www.figma.com/file/aYc5fKU09GawnsQwKTjcDv/Time-bound-stick?node-id=0%3A1&t=mTNGlHiYcWKypsHC-0

AC:
- Flagging a team as exceeding the limits should enable `Usage stats` for all billing leaders
- Enable usage stats only if `Organization` has `activeDomain`
- To enable: 
  - Set `insignts` feature flag for them
  - Set `suggestedTier: 'pro'` for `OrganizationUser` entries 
  (_With suggestedTier=pro, user is only able to see its own organizations in the statistics, which is good. Enabling 'enterprise' without human validation cause security issues, so we are not doing that_)
- Keep Usage stats enabled even if organization upgraded or fixed its limits
- In case Usage Stats enabled because of `limitExceeded`, **do not** show a persistent snackbar: `useUsageSnackNag`
- Instead show one-time closable snackbar:
image

### PR DESCRIPTION
**Pull request by igorlesnenko**
# Description

Some notifications can now be accompanied by a snackbar, because previously they were only displayed as a red dot on top of the bell icon. In the case of lock notifications, we need to show a snackbar to make them more visible

The functionality will also be used for 7 reminder notification

Fixes #7383

Previous PR: #7563

## Demo

Loom: https://www.loom.com/share/ae0023682c0247fcb6fa4219d4361a94

Screenshot 2022-12-20 at 16 00 30

## Testing scenarios

- Go through testing scenario from #7563
- But also see that now a snackbar appears
- Snackbar don't disappear when page is refreshed
- Snackbar disappears when you click on it or read the notification by clicking a bell icon in the header
- You should not see nagging usage stats snackbar in case usage stats were enabled with limit exceeded flag

### CODE REVIEW

**Review by nickoferrall — CHANGES_REQUESTED** [review]
I haven't been able to see the snackbar message in this PR yet. At the moment, it's showing the "View your usage stats" snackbar. Here's a Loom going through the steps I've taken: https://www.loom.com/share/384f3ffd24454c7a96379362f21beda6

**Comment by igorlesnenko** [comment]
@nickoferrall From your video I see that your organization is already flagged so the function that created notification + snackbar didn't run

image

Can you remove `limitExceededAt` flag and try again?

I also see that you have non-closing usage stats snackbar. It should not appear at all when I enable usage stats using my function. So it is probably a bug, if you haven't changed any usage stats related configuration manually. I'll check

**Comment by nickoferrall** [comment]
Thanks @igorlesnenko, I've updated `limitExceededAt` to `null`, but I'm still seeing the same "View your usage stats" snackbar. 

I've tried removing the `insights` feature flag and ran through the testing steps in this Loom, but no luck: https://www.loom.com/share/762cafcccd0a447d9e494c4c9c2d7aa9

**Comment by igorlesnenko** [comment]
@nickoferrall after running the meeting, do you see the flags are now set for the org in the database?

**Comment by nickoferrall** [comment]
@igorlesnenko no they weren't set, they remain `null`

**Comment by igorlesnenko** [comment]
@nickoferrall looks like the function thinks that you didn't exceed the limit. Trying to understand why.

There is a check:

```

  if (teamIds.length <= PERSONAL_TIER_MAX_TEAMS) {
    return false
  }
```

If you have `PERSONAL_TIER_MAX_TEAMS` set to `1` then you need 2 sticky teams for it to work.

Do you have 2 teams?

Can you set `PERSONAL_TIER_MAX_TEAMS` to `0` ?

File: packages/client/components/TopBarNotifications.tsx
Comment (nickoferrall) [inline packages/client/components/TopBarNotifications.tsx:51]: +2 I like this idea. It'll be easy to add further snackbar notifications in the future if we want to

**Review by nickoferrall — CHANGES_REQUESTED** [review]
That's working for me now! I have a second team and thought that I had ran a meeting with it, but perhaps not. 

The "View your usage snackbar" still pops up first and only after I click it away do I see the snackbar added in this PR, so I think we'll need to resolve that before merging this PR?

Also, perhaps it would be useful to comment out the hardcoded values to speed up testing? E.g.

```
const STICKY_TEAM_MIN_MEETING_ATTENDEES = 2
const STICKY_TEAM_MIN_MEETINGS = 3
const PERSONAL_TIER_MAX_TEAMS = 2
const PERSONAL_TIER_LOCK_AFTER_DAYS = 30
// const STICKY_TEAM_MIN_MEETING_ATTENDEES = 1
// const STICKY_TEAM_MIN_MEETINGS = 1
// const PERSONAL_TIER_MAX_TEAMS = 0
// const PERSONAL_TIER_LOCK_AFTER_DAYS = 1
```



Also, perhaps we should

**Comment by jmtaber129** [comment]
Hey @nickoferrall @igorlesnenko! Just popping in to let y'all know that I'm actually proposing a different pattern for notification toasts over in https://github.com/ParabolInc/parabol/pull/7603. 

Also, I noticed that this will emit a snackbar for any unread notifications when the top bar renders, not just newly published notifications (i.e. a notification that just occurred). Is this intentional? This is different from our existing notification patterns. Additionally, if a user signs in with a bunch of notifications implemented in `NotificationSnackbarPicker` (e.g. 10+ instances), the snackbar queue will be pretty full, and newly published notifications won't be seen unless the user dismisses all of previous unread notifications.

**Comment by igorlesnenko** [comment]
@jmtaber129  

>  Is this intentional?

This is intentional.

For my case I need the following behaviour for the snackbar notification:

- Notification could be created when user is not online (e.g. "Your organization is going to be blocked in 30 days!")
- When user opens the app he should see the unread notification as a snackbar
- Notification snackbar should stay across refreshing the page

Would this be possible with your pattern?

**Comment by jmtaber129** [comment]
@igorlesnenko Is there a notion doc or GH epic with additional context I can look at? I feel like I may be missing something here - looking at this PR's issue + Figma file, the snackbar CTA seems like something that would be nagged using `useSnackNag` (similar to usage stats), conditioned on whether the org's `scheduledLockAt` is non-null. As-is, this is depending on the notifications as its source of truth (which feels like a smell), so e.g. one billing leader could upgrade a team, and another billing leader would still see the persistent snackbar notification saying they'll be locked out if their notification is unread.

Regardless, this raises a different tension - IMO implementing this as an abstraction on notifications is in conflict with our existing notification abstractions. Currently, when a notification is created we 1) create + publish the notification to the notification dropdown, and 2) show a toast _if_ the user is online. The way I think about these is that the purpose of (2) is basically to say "hey, this notification _just_ happened" in a transient way, while (1) serves as a more permanent presentation of the notification. This would change (2) to serve the same purpose as (1) for some notifications.

Since this is new logic for a specific notification (which could potentially cause problems if applied to several other notifications, as stated earlier), I'd suggest using something the hook/component for this directly for this notification, rather than implementing this as an abstraction for notifications in general.

**Comment by igorlesnenko** [comment]
@jmtaber129 TLDR: I refactored the code.

> Is there a notion doc or GH epic with additional context I can look at?

Here is the full context:

- https://www.notion.so/parabol/Implement-time-bound-stick-for-enforcing-2-team-limit-in-Usage-Stats-7b6cb3446267449f9269cbc2aa944ae8
- https://www.figma.com/file/aYc5fKU09GawnsQwKTjcDv/Time-bound-stick?node-id=0%3A1&t=VZ8PWu74LBVOy0qD-0
- https://github.com/ParabolInc/parabol/issues/7383

>  the snackbar CTA seems like something that would be nagged using useSnackNag (similar to usage stats)

When using `useSnackNag`, it is not possible to close that snackbar. It will appear every time when user refresh the page. 

We don't want to be that nagging, we want to allow to close the snackbar, but we want it to stay until closed.

So instead of a single nagging snackbar, we want to have a set of notifications — limit exceeded, 14 days left, 7 days left, you're blocked, etc.

This would work with your pattern. 

However, we want the billing leader to see the snackbar anyway, even if he missed it when he wasn't online. 

So when he logins to the app, he should see this notification as a snackbar like if it just arrived.

That what I'm trying to achieve with my PR. If notification are one of the specific types and "unread", show it as a snackbar when page loads.

>  one billing leader could upgrade a team, and another billing leader would still see the persistent snackbar notification saying they'll be locked out if their notification is unread.

That is true, however another billing leader would still see an email saying his organization will be locked. But I think it would be easy to understand that this is a just an "old" message.

However, If we need to handle that case, we can just remove such notifications on organization upgrade (or mark them as "read"). 

> This would change (2) to serve the same purpose as (1) for some notifications.

I would look at it in a different angle. For important notifications we suggest that a small dot near the bell icon have not much visibility, and we want a snackbar to be shown anyway until closed, even if the user was offline. Or different angle — sometimes we just want snackbars to be delivered even if user is offline — "Hey look there was a new notification while you was offline"

---

I refactored my code and now using your approach with a separate component to make `pinned` notifications. (I would still need to improve it a bit)
What do you think?

**Comment by jmtaber129** [comment]
@igorlesnenko thanks for the additional context! I think this makes sense. I do still have some minor forward-looking tensions, but they can be addressed as they come up - these changes are safe to try in my book 👍

File: packages/server/billing/helpers/teamLimitsCheck.ts
Comment (igorlesnenko) [inline packages/server/billing/helpers/teamLimitsCheck.ts:36]: Fixed tier here, that is probably the reason I was not able to see usage stats nagging snackbar

File: packages/client/hooks/useUsageSnackNag.ts
Comment (igorlesnenko) [inline packages/client/hooks/useUsageSnackNag.ts:76]: Fixed usage stats nagging snackbar. Now I don't show it in case limit exceeded flag is set.

**Comment by igorlesnenko** [comment]
@nickoferrall Fixed usage stats snackbar, now it shouldn't appear

File: packages/client/mutations/toasts/mapTeamsLimitExceededToToast.ts
Comment (tghanken) [inline packages/client/mutations/toasts/mapTeamsLimitExceededToToast.ts:32]: +1 I'd like to refactor this event to `Feature CTA Clicked`, with `feature` = "Usage" and `featureCtaLocation` = "Snackbar", but know that it is in use elsewhere in the codebase.

If we decide to merge this as is (which we should do unless we refactor everywhere), I'll open a new issue refactor similar events after I finalize our Segment event naming policy.

**Review by tghanken — APPROVED** [review]

File: packages/client/mutations/toasts/mapTeamsLimitExceededToToast.ts
Comment (igorlesnenko) [reply packages/client/mutations/toasts/mapTeamsLimitExceededToToast.ts:32]: @tghanken I suggest to create a new refactoring issue, to keep this PR small

**Review by nickoferrall — APPROVED** [review]

File: packages/client/components/PinnedSnackbarNotifications.tsx
Comment (Dschoordsch) [inline packages/client/components/PinnedSnackbarNotifications.tsx:74]: -1 I don't understand why this is necessary. It's the same event as calling the action. In fact I see both events are fired here.

File: packages/client/mutations/toasts/popNotificationToast.ts
Comment (Dschoordsch) [inline packages/client/mutations/toasts/popNotificationToast.ts:64]: -1 Why was this `onDismiss` event added? It's also sending the same `'CLICKED'` event as the action.

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Tested, works except for additional mutation calls

File: packages/server/graphql/private/typeDefs/_legacy.graphql
Comment (igorlesnenko) [reply packages/server/graphql/private/typeDefs/_legacy.graphql]: Removed the whole User type

File: packages/client/components/PinnedSnackbarNotifications.tsx
Comment (igorlesnenko) [reply packages/client/components/PinnedSnackbarNotifications.tsx:76]: @Dschoordsch snackbar is attached to a specific notification. We want to mark notification clicked, if snackbar is clicked. However, snackbar could also be dismissed automatically, in this case we don't want `clicked` mutation to be executed.

Refactored code a bit, and added `onManualDismiss` that should cover necessary cases.

File: packages/client/mutations/toasts/popNotificationToast.ts
Comment (Dschoordsch) [inline packages/client/mutations/toasts/popNotificationToast.ts:53]: -1 Why did you remove the action part here? Doesn't that break existing notifications?

**Review by Dschoordsch — CHANGES_REQUESTED** [review]

File: packages/client/components/PinnedSnackbarNotifications.tsx
Comment (igorlesnenko) [reply packages/client/components/PinnedSnackbarNotifications.tsx:76]: @Dschoordsch

>  but there is some precedence in the code where we set CLICKED just to set the notification inactive

That is exactly what I'm trying to do here too.

If the user manually dismiss a snackbar that is attached to a specific notification, I just mark corresponding notification as `clicked`, so it will not appear as unread and will not be highlighted in notifications dropdown.

File: packages/client/mutations/toasts/popNotificationToast.ts
Comment (igorlesnenko) [reply packages/client/mutations/toasts/popNotificationToast.ts:53]: @Dschoordsch this part is not needed anymore. The previous code marked the notification as read only if we pressed the action button. That's why we checked if button exists.

But it could be a snackbar without any button. Or user could just click on other parts of the snackbar to close it.

So now I handle these cases by marking notification as read if snackbar was manually dismissed.

This is the same logic we follow for notifications in header, if you click on it it will be marked as clicked. 

cc @jmtaber129

**Review by Dschoordsch — APPROVED** [review]

**Pull request merged**
